### ISSUE TITLE
Add backpressure support in ledger-api-test-tool

### ISSUE THREAD

**Issue by gerolf-da** [digital-asset/daml#3747]
The `LotsOfParties` test currently runs with a reduced number of parties (125 instead of 1024) because the ledger-api-test-tool doesn't support the backpressure signal on party allocation.

To be able to increase the number of parties in the test again, this support needs to be added to the ledger-api-test-tool.

**Comment by fabiotudone-da** [comment]
Just wanted to add that backpressure might also happen with non-intensive tests.

**Comment by fabiotudone-da** [comment]
I know of some setups that fail quite regularly on `PartyManagementIT` due to `RESOURCE_EXHAUSTED`; FYI @meiersi-da

**Comment by meiersi-da** [comment]
@gerolf-da @stefanobaghino-da - what would it take to make the ledger-api-test-tool work properly with backpressure?

**Comment by stefanobaghino-da** [comment]
If non-intensive tests fail due to resource exhaustion it sounds like a but the test tool should not hide.

**Comment by meiersi-da** [comment]
@stefanobaghino-da -- I'm not sure. In CI resources can be quite constrained. Being able to let the test-tool run with retry-on would be helpful to reduce flakiness. If one were to do that, what bugs would be hidden?

**Comment by stefanobaghino-da** [comment]
Those tests are marked to run exclusively precisely to avoid flakes due to resource exhaustion. There are at most a limited number of concurrent tests running at any given time using the whole CI server (with more intensive tests being run as separate test -- themselves using the CI server exclusively), and back-pressure should kick in when multiple queues with a few hundreds of available slots are filled up. It does not sounds right that in this conditions any of those tests could fail due to a justified and reasonable resource exhaustion.

**Comment by SamirTalwar-DA** [comment]
@stefanobaghino-da: I don't think @meiersi-da is talking about the open-source DAML SDK CI, but rather CI for another, potentially closed-source ledger.

**Comment by fabiotudone-da** [comment]
> If non-intensive tests fail due to resource exhaustion it sounds like a bug the test tool should not hide.

I'd say only if they keep failing, but a transient resource exhaustion condition is not a bug. Also, if the scope of the test is semantic, it should't concern itself with non-functional matters (unless they break the functionality, e.g. totally unavailable).

> I would say that even in the most extreme cases, setting the test tool parallelism to 1 ensures there will be no resource exhaustion.

True, but it can considerably slow down the CI pipeline.

**Comment by stefanobaghino-da** [comment]
> I'd say only if they keep failing, but a transient resource exhaustion condition is not a bug.

If there are only a few tests running at any given time and a few hundreds available queue spots before back-pressure is supposed to kick in, I would say it would be better to investigate what is causing this before hiding it under the rug, not matter how sound implementing back-pressure handling is.

> True, but it can considerably slow down the CI pipeline.

I was giving this as a solution to a CI environment with constrained resources. With constrained resources you are bound to have constrained performances.

**Comment by fabiotudone-da** [comment]
> If there are only a few tests running at any given time and a few hundreds available queue spots before back-pressure is supposed to kick in, I would say it would be better to investigate what is causing this before hiding it under the rug, no matter how sound implementing back-pressure handling is.

Agree, but not by having any functional test fail somehow, rather by having back pressure tests highlight the actual issue.

> I was giving this as a solution to a CI environment with constrained resources. With constrained resources you are bound to have constrained performances.

If you constrain resources artificially in this way though, you could end up under utilizing the available resources. Think of a shared ledger connection pool: you'd need to set the concurrency to match the worst case (probably 1 for 1 connection available in the pool, assuming there must be at least  1 at all times) rather than adopting a more optimistic estimation and just retrying if it fails.

**Comment by stefanobaghino-da** [comment]
We should add back-pressure to the test tool, but I don't think that the recent flakiness should be dealt with by hiding the problem.

There is a known flakiness induced by very low loads since we offloaded DAML-LF value translation off the database threadpool. This is unreasonable in our setup and can be reproduced locally (especially using the H2 conformance tests). There has been some work on this in https://github.com/digital-asset/daml/pull/5983 (making sure we use a work-stealing pool for DAML-LF value translation) but unfortunately this was not enough. I opened https://github.com/digital-asset/daml/issues/6108 to track it.
